flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypointUtils.java

```diff
@@ -128,4 +132,80 @@ public static void configureUncaughtExceptionHandler(Configuration config) {
                 new ClusterUncaughtExceptionHandler(
                         config.get(ClusterOptions.UNCAUGHT_EXCEPTION_HANDLING)));
     }
+
+    /**
+     * Creates the working directory for the TaskManager process. This method ensures that the
+     * working directory exists.
+     *
+     * @param configuration to extract the required settings from
+     * @param resourceId identifying the TaskManager process
+     * @return working directory
+     * @throws IOException if the working directory could not be created
+     */
+    public static WorkingDirectory createTaskManagerWorkingDirectory(
+            Configuration configuration, ResourceID resourceId) throws IOException {
+        return WorkingDirectory.create(
+                generateTaskManagerWorkingDirectoryFile(configuration, resourceId));
+    }
+
+    /**
+     * Generates the working directory {@link File} for the TaskManager process. This method does
+     * not ensure that the working directory exists.
+     *
+     * @param configuration to extract the required settings from
+     * @param resourceId identifying the TaskManager process
+     * @return working directory file
+     */
+    @VisibleForTesting
+    public static File generateTaskManagerWorkingDirectoryFile(
+            Configuration configuration, ResourceID resourceId) {
+        return generateWorkingDirectoryFile(
+                configuration,
+                ClusterOptions.TASK_MANAGER_PROCESS_WORKING_DIR_BASE,
+                "tm_" + resourceId);
+    }
+
+    /**
+     * Generates the working directory {@link File} for the JobManager process. This method does not
+     * ensure that the working directory exists.
+     *
+     * @param configuration to extract the required settings from
+     * @param resourceId identifying the JobManager process
+     * @return working directory file
+     */
+    @VisibleForTesting
+    public static File generateJobManagerWorkingDirectoryFile(
+            Configuration configuration, ResourceID resourceId) {
+        return generateWorkingDirectoryFile(
+                configuration,
+                ClusterOptions.JOB_MANAGER_PROCESS_WORKING_DIR_BASE,
+                "jm_" + resourceId);
+    }
+
+    private static File generateWorkingDirectoryFile(
+            Configuration configuration,
+            ConfigOption<String> precedingOption,
+            String workingDirectoryName) {
+        return new File(
+                configuration
+                        .getOptional(precedingOption)
+                        .orElseGet(
+                                () -> configuration.get(ClusterOptions.PROCESS_WORKING_DIR_BASE)),
```

Comment:
How we handle the case that `PROCESS_WORKING_DIR_BASE` has multi directories from different disks?

Reply:
This case is not supported. The user can only define a single directory for the process working directory base.

Reply:
If so, I think we should add a check here as YARN could set the local dirs with multi disks if possible.

Reply:
At the moment `ClusterOptions.PROCESS_WORKING_DIR_BASE` will default to `System.getProperty("java.io.tmpdir")`. But I think you are right that we should rather default to `CoreOptions.TMP_DIRS` and then make sure that we only take the first returned tmp directory.

Reply:
I've pushed a fixup that changes the behaviour so that `PROCESS_WORKING_DIR_BASE` now defaults to `TMP_DIRS`.

Reply:
After went through current PR, I still have a question that if we always choose the 1st tmp directory as the state access working directory by default. How could we leverage the ability of multi disks in some deployment environment, such as YARN.

How about reshuffle the `defaultDirs` within [BootstrapTools#updateTmpDirectoriesInConfiguration](https://github.com/apache/flink/blob/ed699b6ee6b0539087632b68a444f79b95120d84/flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java#L293-L296), so that we can ensure not return the same tmp directory each time.

Reply:
If users want to use multiple volumes for the RocksDB working directory, then they can set `state.backend.rocksdb.localdir: volumeA,volumeB,volumeC` to use multiple volumes.

If we change `BootstrapTools#updateTmpDirectoriesInConfiguration`, then this should be Yarn specific and should go into some `YarnUtils` class. Or alternatively, the Yarn classes could shuffle the second argument before calling this method.

Reply:
I don't know whether only YARN could make LOCAL dirs as the list of directories from all candidate disks. And this could be fine for all kind of delpoyment systems.

Reply:
But would this really be needed by users? By randomly shuffle the volumes, some processes will fail and other might run if they are lucky and pick a free volume. Do you know of somebody who relies on such a feature?

Reply:
> I don't know whether only YARN could make LOCAL dirs as the list of directories from all candidate disks. And this could be fine for all kind of delpoyment systems.

Hmm, I am a bit hesitant here because it would introduce some magic that is not really easy to understand for users.

Reply:
Since we always get the first volume as the temp working directory, we might also face the problem that the 1st volumn is borken, which leads to no process can run.

We internally migrate to k8s to deploy jobs, but I believe many guys use YARN to deploy jobs on machines with multi disks. I think we should provide friendly user experience under default case.

Reply:
Actually, I am not sure whether we truly support multiple temp directories well enough at this point in time. As a user I would expect that Flink could automatically make use of free directories once others a full. However, this is currently not the case. Flink will simply pick a temp dir for the RocksDB working directory and hope that everything fits in this directory.

Reply:
For a single Flink process, this is ture. However, for many flink processes running on the same machine, it would be very useful to make the usage of disks as much as they can. That's the key point to leverage disk resources for locality applications.

Reply:
Alright. I will introduce this logic and update the config option descriptions accordingly. I would assume that in most cases people would only configure a single temp directory and so it should not make a difference. Moreover, people that want to use the working directory should probably anyway configure an explicit working directory base.